### PR TITLE
Parallelize integration tests

### PR DESCRIPTION
**Pull request by gabesaba**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Speeds up integration tests via parallelization

#### Which issue(s) this PR fixes:
Related to #1737

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -167,7 +167,7 @@ test: gotestsum ## Run tests.
 .PHONY: test-integration
 test-integration: gomod-download envtest ginkgo mpi-operator-crd ray-operator-crd jobset-operator-crd kf-training-operator-crd  cluster-autoscaler-crd ## Run tests.
 	KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" \
-	$(GINKGO) $(GINKGO_ARGS) --junit-report=junit.xml --output-dir=$(ARTIFACTS) -v $(INTEGRATION_TARGET)
+	$(GINKGO) $(GINKGO_ARGS) -procs=4 --junit-report=junit.xml --output-dir=$(ARTIFACTS) -v $(INTEGRATION_TARGET)
```
Comment (tenzen-y) [inline Makefile:170]:
```suggestion
	$(GINKGO) $(GINKGO_ARGS) -procs=$(INTEGRATION_NPROCS) --junit-report=junit.xml --output-dir=$(ARTIFACTS) -v $(INTEGRATION_TARGET)
```

Should we make a parameter for the number of processes?

And then, I think that we can have the `INTEGRATION_NPROCS ?= $(shell nproc)` var.

Comment (trasc) [reply Makefile:170]: According to https://onsi.github.io/ginkgo/#spec-parallelization `-p` will do that.

While running in prow the number of go procs is limited:

https://github.com/kubernetes/test-infra/blob/21fe7ceb7858ecaf0fb9ab38c5222b0b0e2de8bd/config/jobs/kubernetes-sigs/kueue/kueue-presubmits-main.yaml#L50-L60

But if there is chance that 4 resource cpu's can translate in a timeshare of more then 4 we can drop GOMAXPROCS env setting.

Another thing that could be useful is to still be able to run it non-parallel if needed.

Comment (trasc) [inline Makefile:170]: Another thing that could be useful is to still be able to run it non-parallel if needed.

Comment (gabesaba) [reply Makefile:170]: Yes, I'll create a parameter. I prefer to hardcode the value. During testing, I encountered file descriptors exhausting when I set a high-number of processors (see end of [comment](https://github.com/kubernetes-sigs/kueue/issues/1737#issuecomment-1956509955)). I just tested now and I encountered this on my machine when using the value returned by nproc (24).

Comment (gabesaba) [reply Makefile:170]: I added the parameter that @tenzen-y suggested. The user may set the value to 1 to run non-parallel.

Comment (mimowo) [reply Makefile]: I'm ok with making it default to 4 (not the number of processes), due to the issue with file descriptors which would also probably affect my machine. The parameter still allows developers to tune the value if they have more cores.

File: Makefile [hunk 2]
```diff
@@ -71,6 +71,11 @@ E2E_K8S_VERSIONS ?= 1.26.12 1.27.9 1.28.5 1.29.0
 # Default will delete default kind cluster
 KIND_CLUSTER_NAME ?= kind
 
+# Number of processes to use during integration tests to run specs within a
+# suite in parallel. Suites still run sequentially. User may set this value to 1
```
Comment (mimowo) [inline Makefile:75]: "Suites still run sequentially" seems orthogonal to the param. Is there another ginkgo option that let's us control parallel execution of suites?

Comment (gabesaba) [reply Makefile:75]: I don't think so. The [Ginkgo documentation](https://onsi.github.io/ginkgo/#running-multiple-suites) states "when there are multiple suites to run Ginkgo...  always runs them sequentially"

Comment (mimowo) [reply Makefile:75]: Thanks for the ref. In that case I'm ok with  "Suites still run sequentially.".

**Review by mimowo — COMMENTED** [review]
/lgtm
/assign @tenzen-y @alculquicondor

**Comment by mimowo** [comment]
In the "Which issue(s) this PR fixes:" section please also say "Part of #..." or "Related to #". If you write "Fixes #..." the robot closes the issue, we may keep it open to explore other low hanging fruits (hopefully)

**Comment by alculquicondor** [comment]
/approve
/hold
(since we discussed offline that you wanted to double check the times with different values)

Once you are comfortable with the chosen option, you can write `/hold cancel`

File: Makefile (hunk 1 above)
Comment (tenzen-y) [reply Makefile]: Thank you for the clarifications. I'm ok with making it 4 as a default.

**Review by tenzen-y — COMMENTED** [review]
/lgtm

**Comment by gabesaba** [comment]
/hold cancel

INTEGRATION_NPROCS=4 performed better than 2, and 8 failed during a run. Reverting to original value (4).
